File: programmers/js/new/level2/압축.js
function solution(msg) {
    var answer = [];
    const map = new Map();
    for (let i = 1; i <= 26; ++i) {
        map.set(String.fromCharCode("A".charCodeAt(0) + i - 1), i);
    }
    console.log(map);
    let lastIdx = 27;
    for (let i = 0; i < msg.length; ) {
        let str = "";
        let lastStr = "";
        let j = i;
        do {
            lastStr = str;
            str += msg[j++];
        } while (map.has(str));
        console.log(lastStr);
        answer.push(map.get(lastStr));
        map.set(str, lastIdx++);
        i = --j;
    }
    return answer;
}

console.log(solution("KAKAO"));

File: programmers/python/level2/압축.py
def solution(msg):
    answer = []
    dic = {}
    for i, code in enumerate(range(ord('A'), ord('Z')+1)):
        dic[chr(code)] = i + 1
    start_index = 0
    end_index = 0
    while start_index < len(msg):
        end_index = start_index
        current_str = msg[end_index]
        end_index += 1
        while current_str in dic and end_index < len(msg):
            current_str += msg[end_index]
            end_index += 1
        if current_str not in dic:
            answer.append(dic[current_str[0:-1]])
            dic[current_str] = len(dic) + 1
            end_index -= 1
        else:
            answer.append(dic[current_str])
        start_index = end_index
    return answer

print(solution('KAKAO'))